test/e2e/storage/api/testsuite.go

```diff
@@ -0,0 +1,130 @@
+)
+
+// TestSuite represents an interface for a set of tests which works with TestDriver.
+// Each testsuite should implement this interface and use a TestSuiteHandler to define its Tests.
+// RegisterTests function should be called to actually register tests for the TestSuite.
+// All the functions except GetTestSuiteInfo are designed to not be exported so that caller
+// has to use RegisterTests() to register the tests in a standard way.
+type TestSuite interface {
+	GetTestSuiteInfo() TestSuiteInfo
+	// defineTests defines tests of the testpattern for the driver.
+	// Called inside a Ginkgo context that reflects the current driver and test pattern,
+	// so the test suite can define tests directly with ginkgo.It.
+	defineTests(TestDriver, TestPattern)
+	// skipUnsupportedTests will skip the test suite based on the given TestPattern, TestDriver
+	// Testsuite should check if the given pattern and driver works for the "whole testsuite"
+	// Testcase specific check should happen inside defineTests
+	skipUnsupportedTests(TestDriver, TestPattern)
+}
+
+// RegisterTests register the driver + pattern combination to the inside TestSuite
+// This function actually register tests inside testsuite
+func RegisterTests(suite TestSuite, driver TestDriver, pattern TestPattern) {
+	tsInfo := suite.GetTestSuiteInfo()
+	testName := fmt.Sprintf("[Testpattern: %s]%s %s%s", pattern.Name, pattern.FeatureTag, tsInfo.Name, tsInfo.FeatureTag)
+	ginkgo.Context(testName, func() {
+		ginkgo.BeforeEach(func() {
+			// skip all the invalid combination of driver and pattern first
+			SkipInvalidDriverPatternCombination(driver, pattern)
+			// skip the invalid test pattern and driver combination
+			suite.skipUnsupportedTests(driver, pattern)
+		})
+		// actually define the tests
+		// at this step the testsuite should not worry about if the pattern and driver
+		// does not fit for the whole testsuite. But driver&pattern check
+		// might still needed for specific independent test cases.
+		suite.defineTests(driver, pattern)
+	})
+}
+
+// DefineTestSuites defines tests for all testpatterns and all testSuites for a driver
+func DefineTestSuites(driver TestDriver, tsInits []func() TestSuite) {
+	for _, testSuiteInit := range tsInits {
+		suite := testSuiteInit()
+		for _, pattern := range suite.GetTestSuiteInfo().TestPatterns {
+			RegisterTests(suite, driver, pattern)
+		}
+	}
+}
+
+// TestSuiteInfo represents a set of parameters for TestSuite
+type TestSuiteInfo struct {
+	Name               string              // name of the TestSuite
+	FeatureTag         string              // featureTag for the TestSuite
+	TestPatterns       []TestPattern       // Slice of TestPattern for the TestSuite
+	SupportedSizeRange e2evolume.SizeRange // Size range supported by the test suite
+}
+
+// SkipInvalidDriverPatternCombination will skip tests if the combination of driver, and testpattern
+// is not compatible to be tested. This function will be called in the TestSuiteHandler to make
```

Comment:
there is no TestSuiteHandler, right?

Reply:
Yep, that is removed per @chrishenzie suggest.